ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp

```diff
@@ -0,0 +1,160 @@
+        if (rm_tensor.is_contiguous()) {
+            // Page size depends on the width, so only modify the shape if the width is the same
+            if (tensor_shape.with_tile_padding()[-1] == shape.with_tile_padding()[-1]) {
+                return rm_tensor.reshape(shape.value);
+            }
+            //Different page width, going to use device kernel that does transpose
+            else {
+                auto original_rank = shape.rank();
+                auto tensor_4d = unsqueeze_to_4D(rm_tensor);
+                const auto shape_4d = shape.to_rank<4>();
+                auto reshaped_tensor = ttnn::reshape_on_device(tensor_4d, shape_4d[0], shape_4d[1], shape_4d[2], shape_4d[3], tensor.memory_config());
+                reshaped_rm_tensor = squeeze_from_4D(reshaped_tensor, original_rank);
+            }
+        } else if (tensor_shape.rank() >= 2 and shape.rank() >= 2) {
+            // Handle the case when the tensor is not contiguous but the last two dimensions are the same and so reshape
+            // is possible
+            if (tensor_shape[-1] == shape[-1] and tensor_shape[-2] == shape[-2] and
+                tensor_shape.with_tile_padding()[-1] == shape.with_tile_padding()[-1] and
+                tensor_shape.with_tile_padding()[-2] == shape.with_tile_padding()[-2]) {
+                reshaped_rm_tensor = rm_tensor.reshape(shape.value);
+            }
+        } else {
+            reshaped_rm_tensor = host_reshape(tensor, shape);
+        }
+
+    }
+    // Can'd do untilize on device due to inner dim size
+    else {
+        reshaped_rm_tensor = host_reshape(tensor, shape);
+    }
+
+    if (((shape[-1] * tensor.element_size()) % sizeof(uint32_t) == 0) and reshaped_rm_tensor.layout() != layout) {
+        return ttnn::to_layout(reshaped_rm_tensor, layout, std::nullopt, std::nullopt, (Device *)nullptr);
+    }
+    else {
+        return reshaped_rm_tensor;
+    }
+
+}
+
+ttnn::Shape get_shape_from_vector(const ttnn::Tensor& tensor, const std::vector<int32_t> & shape) {
+    std::int64_t new_volume = 1;
+    std::int64_t index_of_negative_1 = -1;
+
+    for (auto index = 0; index < shape.size(); ++index) {
+        if (shape[index] == -1) {
+            if (index_of_negative_1 != -1) {
+                TT_THROW("Shape cannot have more than 1 elements that is set to -1!");
+            }
+            index_of_negative_1 = index;
+        }
+        new_volume *= shape[index];
+    }
+
+    std::vector<std::uint32_t> new_shape(shape.size());
+    std::copy(shape.begin(), shape.end(), new_shape.begin());
+    if (new_volume < 0) {
+        const auto volume = tensor.get_shape().with_tile_padding().volume();
+        new_shape[index_of_negative_1] = volume / (-new_volume);
+    }
+    return ttnn::Shape(new_shape);
+}
+
+}
+
+
+ttnn::Tensor ReshapeViewOperation::invoke(
+    const ttnn::Tensor& tensor,
+    const ttnn::Shape& shape
```

Comment:
Why don't you call `detail::get_shape_from_vector(tensor, shape_vector)` here?

Reply:
I think ttnn::Shape cannot contain negatives as the underlying LegacyShape that it uses for its value uses std::array<uint32_t, MAX_NUM_DIMENSIONS>